src/main/kotlin/com/nikitanazarov/coroutinestacks/CoroutineStacksPanel.kt

```diff
@@ -43,22 +53,274 @@ class CoroutineStacksPanel(project: Project) : JBPanelWithEmptyText() {
+            parallelStackWindow?.add(i)
+            parallelStackWindow?.add(Box.createVerticalStrut(25))
+        }
+
+    }
+
+    private fun addCoroutineInfoToParallelStackWindow(
+        node: CoroutineStacksPanel.TreeNode<CoroutineStacksPanel.ParallelStackNode>?,
+        rows: MutableList<Box>,
+        level: Int) {
+
+        if (node == null) {
+            return
+        }
+
+        if (node.value.stacktrace.isNotEmpty()) {
+            addCoroutineInfoBox(node.value, rows[level])
+        }
+        for (child in node.children) {
+            addCoroutineInfoToParallelStackWindow(child, rows,level + 1)
+        }
+
+    }
+
+    private fun addCoroutineInfoBox(
+        value: CoroutineStacksPanel.ParallelStackNode,
+        box: Box) {
+
+        val headerText = "${value.additionalData.size} Coroutines"
+        val coroutineList = mutableListOf<String>()
+        coroutineList.add(headerText)
+
+        for (i in value.stacktrace) {
+            i?.let { coroutineList.add(it) }
+        }
+
+        val coroutineListView = JBList<String>(coroutineList)
+
+        coroutineListView.cellRenderer = SeparatedListCellRenderer()
+        val scrollPane = JBScrollPane(coroutineListView)
+
+        val border = BorderFactory.createLineBorder(Color.BLACK, 1)
+        scrollPane.border = border
+
+        box.add(scrollPane)
+        box.add(Box.createHorizontalStrut(120))
+
+    }
+
+    internal class SeparatedListCellRenderer : DefaultListCellRenderer() {
+        override fun getListCellRendererComponent(
+            list: JList<*>,
+            value: Any,
+            index: Int,
+            isSelected: Boolean,
+            cellHasFocus: Boolean
+        ): Component {
+            val renderer = super.getListCellRendererComponent(list, value, index, isSelected, cellHasFocus)
+            if (index < list.model.size - 1) {
+                (renderer as JComponent).border = ITEM_BORDER
+            } else {
+                (renderer as JComponent).border = null
+            }
+            return renderer
+        }
+
+        companion object {
+            private val ITEM_BORDER = BorderFactory.createMatteBorder(0, 0, 1, 0, Color.GRAY)
+        }
+    }
+
+    private fun generateParallelStackTree(
+        tree: CoroutineStacksPanel.Tree<CoroutineStacksPanel.ParallelStackNode>,
+        rootValue: CoroutineStacksPanel.ParallelStackNode?,
+        positionOfStackFrame: Int) {
+
+        val idToStackFramesMap = mutableMapOf<CoroutineInfoData, String>()
+        if (rootValue != null) {
+            for (k in rootValue.additionalData!!) {
+                if (k.stackTrace.size > positionOfStackFrame)
+                    idToStackFramesMap[k] = k.stackTrace[k.stackTrace.size -1 - positionOfStackFrame].toString()
+            }
+        }
+
+        println("idToStackFramesMap: $idToStackFramesMap")
+
+        val groupedPositions = idToStackFramesMap.entries.groupBy { it.value }.values.map { list ->
```

Comment:
This line is very hard to read, could you please split the call chain to make it more readable?